### PR TITLE
59/swteam/openid linking

### PR DESCRIPTION
**Pull request by fpottbaecker**
This will add the ability to link existing accounts with OpenID and to sign in using OpenID when the account is linked.

Since this almost 1k changes PR might be a little scary at first, here's a list of interesting files:

* [omniauth_callbacks_controller.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-a4275c3adf0e6810f2de8c6b215f6645)
* [users_controller.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-4e05ad0d64e6100656b63ad1e78f32c5)
* [user.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-4676c008b11a5480d73d4a6de01e45b9)
* [edit.html.erb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-9ebe6f1b284635a3978b13e45dd525f3)
* [new.html.erb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-ff9301c8a29bd406e850e084bc89a8a7)
* [devise.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-a8a4112d0760c0fd11f691c6a5fb22ad)
* [devise.en.yml](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-f3d6e5d1f2709fef60571f46916a5f23)
* [routes.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-21497849d8f00507c9c8dcaf6288b136)
* [20171119145250_add_omniauth_to_users.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-c3f6dcc577d8675931053213af30a2dd)
* [20171121154832_add_unique_omniauth_to_users.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-8d15fb45e84efa468c8e5d7798d760c7)
* [linking.feature](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-6a5eb2af54147d71f53f22859f5ca82e)
* [omniauth_callbacks_controller_spec.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-48caf2ed2ef2ff70a302c60854b1566e)
* [users_controller_spec.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-289c53b4c0f454861db9a893a5ceb957)
* [user_spec.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-12b107c16792b9ecba685e51b51826f1)

If you are very interested, these files might be relevant as well:

* Cucumber step definitions [this](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-eb4909f65a59b04c5f2d241ff2de9693) and below, specifically [env.local.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-34ebe88fa12d4522140e38cd896f0993)
* [Gemfile](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-8b7db4d5cc4b8f6dc8feb7030baa2478)
* [schema.rb](https://github.com/hpi-swt2/sport-portal/pull/90/files#diff-1acd2e7e27a227829d5d14a91c863bb6)

### CODE REVIEW

**Comment by fpottbaecker** [comment]
I know this is a very big PR, i have included a list of interesting files in the description, could you please review it, especially @Markus-Brand and @ErikBrendel?

**Review by MerlindlH — APPROVED** [review]
Acceptance tests and manual testing look good

File: config/locales/devise.en.yml [hunk 1]
```diff
@@ -28,8 +28,13 @@ en:
...
     omniauth_callbacks:
-      failure: "Could not authenticate you from %{kind} because \"%{reason}\"."
-      success: "Successfully authenticated from %{kind} account."
+      user:
+        link_success: "Successfully connected with %{provider}."
+        link_failure: "Failed to connect with %{provider}, you are already using another provider."
+        link_taken: "Failed to connect with %{provider}, someone is already linked to that account."
+        success: "Successfully authenticated with %{provider}."
+        failure: "Could not authenticate with %{provider}, the account is not linked."
+      failure: "Could not authenticate with %{provider} because \"${reason}\"."
```
Comment (ErikBrendel) [inline config/locales/devise.en.yml]: `$`->`%`

File: features/step_definitions/omniauth.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+Given(/^an unlinked OpenID account$/) do
+  create_account
+end
+
+Given(/^the account is used for authentication$/) do
+  authenticate_with single_account
+end
+
```
Comment (ErikBrendel) [inline features/step_definitions/omniauth.rb:8]: spaaaaace

File: features/step_definitions/omniauth.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+Given(/^an unlinked OpenID account$/) do
+  create_account
```
Comment (ErikBrendel) [inline features/step_definitions/omniauth.rb]: please rename (to make clear that you create an openId account here)

File: features/step_definitions/user.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+end
+
+Then(/^(\w+) should not be linked with any account$/) do |name|
+  user = user_named(name)
+  user.reload
+  expect(user.uid).to be_nil
+  expect(user.provider).to be_nil
+end
+
+
```
Comment (ErikBrendel) [inline features/step_definitions/user.rb]: .

File: spec/controllers/users/omniauth_callbacks_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
+require 'rails_helper'
+
+RSpec.describe Users::OmniauthCallbacksController, type: :controller do
+  describe 'GET #hpiopenid' do
+    before :each do
+      @request.env['devise.mapping'] = Devise.mappings[:user]
+      @user = FactoryBot.create(:user, uid: '1234567890', provider: 'hpiopenid')
+      @autohash = OmniAuth::AuthHash.new(
```
Comment (ErikBrendel) [inline spec/controllers/users/omniauth_callbacks_controller_spec.rb]: ohhhhh, 🚗

File: spec/controllers/users/omniauth_callbacks_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        sign_in @user2
+        get :hpiopenid
+        @user.reload
+        expect(response).to redirect_to(user_path(@user2))
+        expect(@user2.uid).to be_nil
+        expect(@user2.provider).to be_nil
+      end
+    end
+
+    context 'given an linked, logged in user' do
```
Comment (ErikBrendel) [inline spec/controllers/users/omniauth_callbacks_controller_spec.rb]: an

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -17,8 +17,72 @@
...
+    end
+  end
+
+  describe '#has_omniauth' do
+    it 'should return return true iff the user has an omniauth provider and uid' do
+      user = FactoryBot.create :user, provider: 'mock', uid: '1234567890'
+      expect(user.has_omniauth?).to be(true)
+    end
+
+    it 'should return return false if the user does not have a provider or an uid' do
```
Comment (ErikBrendel) [inline spec/models/user_spec.rb]: an?

**Review by ErikBrendel — CHANGES_REQUESTED** [review]
you could have separated cucumber from your features for better PR readability.

File: features/step_definitions/omniauth.rb (hunk 2 above)
Comment (fpottbaecker) [reply features/step_definitions/omniauth.rb:8]: done

**Review by ErikBrendel — APPROVED** [review]
👍

**Pull request merged**